# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/WebrtcGmpVideoCodec.cpp
# Commit: 9c021d42060a
# Full Hash: 9c021d42060a464f5c1acbf917d8280a569dc6e9
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2025-03-14 04:45:14
# Description:
#   Bug 1953722 - Key input frames to the webrtc gmp encoder off ntp timestamps instead of rtp. r=aosmond
#   
#   This to avoid time going backwards due to uint32_t wraparound.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D241394
# ==============================================================================

diff -r d0045739b9f8 -r 9c021d42060a dom/media/webrtc/libwebrtcglue/WebrtcGmpVideoCodec.cpp
--- a/dom/media/webrtc/libwebrtcglue/WebrtcGmpVideoCodec.cpp	Thu Mar 13 15:11:44 2025 +0000
+++ b/dom/media/webrtc/libwebrtcglue/WebrtcGmpVideoCodec.cpp	Thu Mar 13 15:25:10 2025 +0000
@@ -387,8 +387,7 @@
     GMP_LOG_DEBUG("GMP Encode: failed to create frame");
     return;
   }
-  frame->SetTimestamp((aInputImage.rtp_timestamp() * 1000ll) /
-                      90);  // note: rounds down!
+  frame->SetTimestamp(AssertedCast<uint64_t>(aInputImage.ntp_time_ms() * 1000));
 
   GMPCodecSpecificInfo info{};
   info.mCodecType = kGMPVideoCodecH264;
@@ -420,10 +419,14 @@
   MOZ_ASSERT(frameConfigs.size() == 1);
 
   MOZ_RELEASE_ASSERT(mInputImageMap.IsEmpty() ||
-                     mInputImageMap.LastElement().rtp_timestamp <
-                         frame->Timestamp());
-  mInputImageMap.AppendElement(InputImageData{
-      frame->Timestamp(), aInputImage.timestamp_us(), frameConfigs[0]});
+                     mInputImageMap.LastElement().ntp_timestamp_ms <
+                         aInputImage.ntp_time_ms());
+  mInputImageMap.AppendElement(
+      InputImageData{.gmp_timestamp_us = frame->Timestamp(),
+                     .ntp_timestamp_ms = aInputImage.ntp_time_ms(),
+                     .timestamp_us = aInputImage.timestamp_us(),
+                     .rtp_timestamp = aInputImage.rtp_timestamp(),
+                     .frame_config = frameConfigs[0]});
 
   GMP_LOG_DEBUG("GMP Encode: %" PRIu64, (frame->Timestamp()));
   err = mGMP->Encode(std::move(frame), codecSpecificInfo, gmp_frame_types);
@@ -526,18 +529,19 @@
     const nsTArray<uint8_t>& aCodecSpecificInfo) {
   MOZ_ASSERT(mGMPThread->IsOnCurrentThread());
   Maybe<InputImageData> data;
-  auto rtp_comparator = [](const InputImageData& aA,
-                           const InputImageData& aB) -> int32_t {
-    const auto& a = aA.rtp_timestamp;
-    const auto& b = aB.rtp_timestamp;
+  auto gmp_timestamp_comparator = [](const InputImageData& aA,
+                                     const InputImageData& aB) -> int32_t {
+    const auto& a = aA.gmp_timestamp_us;
+    const auto& b = aB.gmp_timestamp_us;
     return a < b ? -1 : a != b;
   };
   size_t nextIdx = mInputImageMap.IndexOfFirstElementGt(
-      InputImageData{aEncodedFrame->TimeStamp(), 0}, rtp_comparator);
+      InputImageData{.gmp_timestamp_us = aEncodedFrame->TimeStamp()},
+      gmp_timestamp_comparator);
   const size_t numToRemove = nextIdx;
   size_t numFramesDropped = numToRemove;
   MOZ_ASSERT(nextIdx != 0);
-  if (nextIdx != 0 && mInputImageMap.ElementAt(nextIdx - 1).rtp_timestamp ==
+  if (nextIdx != 0 && mInputImageMap.ElementAt(nextIdx - 1).gmp_timestamp_us ==
                           aEncodedFrame->TimeStamp()) {
     --numFramesDropped;
     data = Some(mInputImageMap.ElementAt(nextIdx - 1));
@@ -583,7 +587,6 @@
 
   webrtc::VideoFrameType ft;
   GmpFrameTypeToWebrtcFrameType(aEncodedFrame->FrameType(), &ft);
-  uint64_t timestamp = (aEncodedFrame->TimeStamp() * 90ll + 999) / 1000;
 
   GMP_LOG_DEBUG("GMP Encoded: %" PRIu64 ", type %d, len %d",
                 aEncodedFrame->TimeStamp(), aEncodedFrame->BufferType(),
@@ -600,8 +603,9 @@
   unit.SetEncodedData(webrtc::EncodedImageBuffer::Create(
       aEncodedFrame->Buffer(), aEncodedFrame->Size()));
   unit._frameType = ft;
-  unit.SetRtpTimestamp(timestamp);
+  unit.SetRtpTimestamp(data->rtp_timestamp);
   unit.capture_time_ms_ = webrtc::Timestamp::Micros(data->timestamp_us).ms();
+  unit.ntp_time_ms_ = data->ntp_timestamp_ms;
   unit._encodedWidth = aEncodedFrame->EncodedWidth();
   unit._encodedHeight = aEncodedFrame->EncodedHeight();
 